### PR TITLE
After starting running, keep sending keep_alive until receive stop

### PR DESCRIPTION
**Pull request by MateusPedrosa**
Depends on:
- [ ] https://github.com/tidewise/drivers-genset_whisperpower_ddc/pull/4

- Change implementation to follow the prinple used in
drivers/genset_whisperpower_ddc/src/Main.cpp. If receive start command
on the input port, start the generator and then keep sending keep_alive
command every time a new valid frame is received.
- Stop checking the command on the received frame to consider it valid.
- Add respective tests.

### CODE REVIEW

File: tasks/Task.cpp [hunk 1]
```diff
@@ -60,8 +60,8 @@ void Task::updateHook()
 
 void Task::processIO()
 {
+    ControlCommand sentCommand = CONTROL_CMD_NO_COMMAND;
```
Comment (doudou) [inline tasks/Task.cpp]: Declare local variables close to where they are to be used (i.e. just before the `if(m_running)`)

File: tasks/Task.hpp [hunk 2]
```diff
@@ -33,6 +33,8 @@ tasks/Task.cpp, and will be put in the genset_whisperpower_ddc namespace.
 
         std::unique_ptr<VariableSpeedMaster> m_driver;
 
+        bool m_running = false;
```
Comment (doudou) [inline tasks/Task.hpp]: Because there is a manual start button, we should initialize m_running on start, or the software might shut down a running generator on reboot.

File: tasks/Task.cpp [hunk 3]
```diff
@@ -78,25 +78,34 @@ void Task::processIO()
...
-     * Only send control command after receiving a valid frame of an expected type coming from the generator
-     * If received frame is valid but not of an expected type or comes from another source, just ignore it and
+     * Generator starts running when it receives the start command. After that, keep sending the keep_alive command to the generator
+     * until receive stop control command on the input port.
+     * Only send control command after receiving a valid frame coming from the generator.
+     * If received frame is valid but comes from another source or has a different target, just ignore it and
      * don't send control command
      */
-    if (frame.targetID == variable_speed::TARGET_ADDRESS && frame.sourceID == variable_speed::SOURCE_ADDRESS) {
+    if (frame.targetID == variable_speed::PANELS_ADDRESS && frame.sourceID == variable_speed::DDC_CONTROLLER_ADDRESS) {
```
Comment (doudou) [inline tasks/Task.cpp]: Avoid deeply nested ifs. Return early with

~~~
if (frame.targetID != ... || frame.sourceID != ...) {
    return;
}
~~~

File: tasks/Task.cpp [hunk 4]
```diff
@@ -78,25 +78,34 @@ void Task::processIO()
...
         else if (frame.command == variable_speed::PACKET_RUN_TIME_STATE) {
             _run_time_state.write(m_driver->parseRunTimeState(frame.payload, now));
-
-            while (_control_cmd.read(m_control_cmd) == RTT::NewData) {
-                m_driver->sendControlCommand(m_control_cmd);
+        }
+        if (m_running) {
+            sentCommand = CONTROL_CMD_KEEP_ALIVE;
+        }
+        while (_control_cmd.read(m_control_cmd) == RTT::NewData) {
```
Comment (doudou) [inline tasks/Task.cpp]: You don't need both m_control_cmd and m_running. Actually, I think you should just keep m_control_cmd (renamed to m_last_command).

We read in attributes when the read value needs to be reused, or in critical codepaths to avoid memory allocations. But the control command has no allocation so you could have defined it as a local variable. 

Now, moving this logic to a separate method and removing m_running would dramatically reduce the number of branches (a good metric for "simplify"). In particular, it avoids the sentCommand temporary, which is really hard to follow:

In `updateHook`:

~~~
m_last_cmd = processStartStopCommand();
~~~

The new method:

~~~
ControlCommand Task::processStartStopCommand() {
    ControlCommand cmd;
    if (_control_cmd.read(cmd) == RTT::NoData) {
        return m_last_cmd;
    }

    if (cmd != m_last_cmd) {
        m_driver->sendControlCommand(cmd);
        return cmd
    }

    if (m_last_cmd == CONTROL_CMD_START) {
        m_driver->sendControlCommand(CONTROL_CMD_KEEPALIVE);
    }
    return m_last_cmd;
}
~~~

**Comment by doudou** [comment]
There is one last bit, that I missed in the first review, but is now very obvious:

Using `ControlCommand` as the type of the input port does not make sense anymore. It has 4 different values, two of which we don't process, and one of them (`_START`) is actually interpreted by the component to mean something else (_KEEPALIVE).

Just make it a boolean.

File: tasks/Task.cpp [hunk 5]
```diff
@@ -50,6 +51,9 @@ bool Task::startHook()
 {
     if (! TaskBase::startHook())
         return false;
+
+    m_running = false;
```
Comment (doudou) [inline tasks/Task.cpp]: You are still assuming that the command on start is "stop". This is wrong, as if for whatever reason the software needs to reboot, it will shut down the generator (assume a faulty battery, what happens then ?).

The component needs to keep the state of the generator "as it found it on start".

What would be reasonable doing is, in startHook, to wait receiving the generator state and initialize m_running using that information.

File: test/task_test.rb [hunk 6]
```diff
@@ -47,25 +39,25 @@ def iodrivers_base_prepare(model)
...
-                     )
-                 end
-                 .to { have_one_new_sample task.generator_state_port }
+        sample = assert_driver_processes_frame(frame, task.generator_state_port)
 
-        assert_equal (((2 * Math::PI) / 60) * ((0x01 << 8) | 0x00)).round(5),
-                     sample.rotation_speed.round(5)
-        assert_equal (0.01 * ((0x03 << 8) | 0x02)).round(5),
-                     sample.start_battery_voltage.round(5)
+        assert_equal(
```
Comment (doudou) [inline test/task_test.rb]: There is `assert_in_delta` for this. `assert_in_delta x, y, precision`. You usually can omit the precision (the default is good for a lot of common cases)

Every test harness / framework has special assertions for floating-point comparison.

**Review by doudou — APPROVED** [review]

File: tasks/Task.cpp [hunk 7]
```diff
@@ -119,4 +134,63 @@ void Task::cleanupHook()
...
\ No newline at end of file
+}
+
+bool Task::processStartStopCommand()
+{
+    bool cmd;
+    if (_control_cmd.read(cmd) == RTT::NoData) {
+        if (m_running) {
+            m_driver->sendControlCommand(CONTROL_CMD_KEEP_ALIVE);
+        }
```
Comment (doudou) [inline tasks/Task.cpp]: `NoData` should really mean "do not do anything"

File: test/task_test.rb [hunk 8]
```diff
@@ -15,6 +15,7 @@
                   .deployed_as("genset_whisperpower_ddc")
         )
         @task.properties.io_read_timeout = Time.at(2)
+        self.expect_execution_default_timeout = 30
```
Comment (doudou) [inline test/task_test.rb]: Do we really need this ?

File: tasks/Task.cpp (hunk 7 above)
Comment (MateusPedrosa) [reply tasks/Task.cpp]: The `control_cmd` port receives true or false to respectively start or stop the generator. After getting it running, the component should repeatedly send keep_alive even if there isn't anything on this input port, otherwise we would go back to having to manually send the keep_alive command every time.

Comment (MateusPedrosa) [reply tasks/Task.cpp]: Maybe you prefer to continuously send true to keep it running?

Comment (doudou) [reply tasks/Task.cpp]: Ah ... misunderstanding here.

`NoData` means nothing has ever been written on the port (or .clear() was called). When you read a port for which there has been some data written, you get OldData (and the value is re-read in the argument)

File: tasks/Task.cpp [hunk 9]
```diff
@@ -59,6 +60,8 @@ bool Task::startHook()
     if (! TaskBase::startHook())
         return false;
 
+    std::cout << "RUNNING: " << endl;
+
```
Comment (doudou) [inline tasks/Task.cpp]: Leftover debug

File: tasks/Task.cpp [hunk 10]
```diff
@@ -4,6 +4,7 @@
 #include <iodrivers_base/ConfigureGuard.hpp>
 #include <genset_whisperpower_ddc/VariableSpeed.hpp>
 #include <genset_whisperpower_ddc/ControlCommand.hpp>
+#include <fstream>
```
Comment (doudou) [inline tasks/Task.cpp]: Also leftover because of the cout ? You only need iostream for cout

Comment (MateusPedrosa) [reply tasks/Task.cpp]: Sorry, should have paid more attention. It was actually for a workaround I did before you showed me the correct location of the log file.

**Review by doudou — APPROVED** [review]

**Comment by doudou** [comment]
Thanks @MateusPedrosa. I'm going to merge it myself, I'd like to have a chance to have it built for tomorrow morning.
